semantic-core/src/Control/Effect/Readline.hs

```diff
@@ -1,108 +1,30 @@
-import Control.Monad.IO.Class
-import Control.Monad.Trans.Class
-import Data.Int
-import Data.String
 import Data.Text.Prettyprint.Doc
-import Data.Text.Prettyprint.Doc.Render.Text
+import Data.Text.Prettyprint.Doc.Render.Terminal
 import GHC.Generics (Generic1)
-import System.Console.Haskeline hiding (Handler, handle)
-import System.Directory
-import System.FilePath
+import Prelude hiding (print)
 
 data Readline m k
-  = Prompt String (Maybe String -> m k)
-  | Print AnyDoc (m k)
-  | AskLine (Line -> m k)
-  deriving stock (Functor, Generic1)
-  deriving anyclass (Effect, HFunctor)
-
-newtype AnyDoc = AnyDoc { unAnyDoc :: forall a . Doc a }
-
-prompt :: (IsString str, Member Readline sig, Carrier sig m) => String -> m (Maybe str)
-prompt p = fmap fromString <$> send (Prompt p pure)
-
-print :: (Pretty a, Carrier sig m, Member Readline sig) => a -> m ()
-print s = send (Print (AnyDoc (pretty s)) (pure ()))
-
-println :: (Pretty a, Carrier sig m, Member Readline sig) => a -> m ()
-println s = print s >> print @String "\n"
-
-askLine :: (Carrier sig m, Member Readline sig) => m Line
-askLine = send (AskLine pure)
-
-newtype Line = Line Int64
-
-increment :: Line -> Line
-increment (Line n) = Line (n + 1)
-
-newtype ReadlineC m a = ReadlineC { runReadlineC :: ReaderC Line (LiftC (InputT m)) a }
-  deriving newtype (Applicative, Functor, Monad, MonadIO)
-
-runReadline :: MonadException m => Prefs -> Settings m -> ReadlineC m a -> m a
-runReadline prefs settings = runInputTWithPrefs prefs settings . runM . runReader (Line 0) . runReadlineC
-
-instance (MonadException m, MonadIO m) => Carrier (Readline :+: Lift (InputT m)) (ReadlineC m) where
-  eff (L (Prompt prompt k)) = ReadlineC $ do
-    str <- lift (lift (getInputLine (cyan <> prompt <> plain)))
-    local increment (runReadlineC (k str))
-    where cyan = "\ESC[1;36m\STX"
-          plain = "\ESC[0m\STX"
-  eff (L (Print text k)) = liftIO (putDoc (unAnyDoc text)) *> k
-  eff (L (AskLine k)) = ReadlineC ask >>= k
-  eff (R other) = ReadlineC (eff (R (handleCoercible other)))
-
-runReadlineWithHistory :: MonadException m => ReadlineC m a -> m a
-runReadlineWithHistory block = do
-  homeDir <- liftIO getHomeDirectory
-  prefs <- liftIO $ readPrefs (homeDir </> ".haskeline")
-  let settingsDir = homeDir </> ".local/semantic-core"
-      settings = Settings
-        { complete = noCompletion
-        , historyFile = Just (settingsDir <> "/repl_history")
-        , autoAddHistory = True
-        }
-  liftIO $ createDirectoryIfMissing True settingsDir
-
-  runReadline prefs settings block
-
-runControlIO :: (forall x . m x -> IO x) -> ControlIOC m a -> m a
-runControlIO handler = runReader (Handler handler) . runControlIOC
-
--- | This exists to work around the 'MonadException' constraint that haskeline entails.
-newtype ControlIOC m a = ControlIOC { runControlIOC :: ReaderC (Handler m) m a }
-  deriving newtype (Applicative, Functor, Monad, MonadIO)
+  = Prompt String (Int -> Maybe String -> m k)
```

Comment:
`Line` is now an implementation detail of the carrier; plus we now return the line number along with the text (if any) from the `Prompt` constructor, and 🔥 `AskLine`.